File: src/main/java/com/smarthome/backend/server/api/modules/hue/HueTemperatureSensor.java
package com.smarthome.backend.server.api.modules.hue;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.smarthome.backend.model.devices.DeviceTemperature;
import com.smarthome.backend.server.api.modules.Module;
/**
 * Repräsentiert einen Hue Temperature Sensor als Gerät.
 * Erweitert DeviceTemperature und implementiert die Hue-spezifische Kommunikation.
 */
public class HueTemperatureSensor extends DeviceTemperature {
    private static final Logger logger = LoggerFactory.getLogger(HueTemperatureSensor.class);
    
    private String bridgeId;
    private String hueResourceId; // Die ID des Temperature-Sensors in der Hue Bridge
    private String batteryRid;
    @com.google.gson.annotations.Expose(serialize = false, deserialize = false)
    private transient HueDeviceController hueDeviceController;
    /**
     * Standardkonstruktor für Gson-Deserialisierung.
     * Wird verwendet, wenn Geräte aus der Datenbank geladen werden.
     */
    public HueTemperatureSensor() {
        super();
        this.bridgeId = null;
        this.hueResourceId = null;
        this.batteryRid = null;
        this.hueDeviceController = null;
        setModuleId(Module.HUE.getModuleId());
        setIsConnected(true);
    }
    
    /**
     * Erstellt einen neuen HueTemperatureSensor.
     * 
     * @param name Der Name des Temperature-Sensors
     * @param id Die eindeutige ID des Temperature-Sensors
     * @param bridgeId Die ID der Hue Bridge, zu der dieser Sensor gehört
     * @param hueResourceId Die ID des Temperature-Sensors in der Hue Bridge
     * @param batteryRid Die ID der Batterie des Temperature-Sensors
     */
    public HueTemperatureSensor(String name, String id, String bridgeId, String hueResourceId, String batteryRid, HueDeviceController hueDeviceController) {
        super(name, id);
        this.bridgeId = bridgeId;
        this.hueResourceId = hueResourceId;
        this.batteryRid = batteryRid;
        this.hueDeviceController = hueDeviceController;
        setModuleId(Module.HUE.getModuleId());
        setIsConnected(true);
    }
    
    @Override
    public void updateValues() {
        logger.info("Update die Werte für {} ", getId());
        if (this.hueDeviceController == null) {
            logger.warn("HueDeviceController ist null - kann Temperaturwert nicht initialisieren für {}", getId());
            return;
        }
        this.hueDeviceController.getTemperature(this).thenAccept(temperature -> {
            super.temperature = temperature;
        }).exceptionally(e -> {
            logger.error("Fehler beim Initialisieren der Temperaturwert für {}", getId(), e);
            return null;    
        });// Standard-Temperatur in Grad Celsius
    }
    
    @Override
    protected void executeSetTemperature(int temperature) {}
    
    public String getBridgeId() {
        return bridgeId;
    }
    
    public void setBridgeId(String bridgeId) {
        this.bridgeId = bridgeId;
    }
    
    public String getHueResourceId() {
        return hueResourceId;
    }
    
    public void setHueResourceId(String hueResourceId) {
        this.hueResourceId = hueResourceId;
    }

    public String getBatteryRid() {
        return batteryRid;
    }

    public void setBatteryRid(String batteryRid) {
        this.batteryRid = batteryRid;
    }
    
    /**
     * Setzt den HueDeviceController für diesen Sensor.
     * Wird verwendet, wenn der Sensor aus der Datenbank geladen wird und der Controller später gesetzt werden muss.
     * 
     * @param hueDeviceController Die HueDeviceController-Instanz
     */
    public void setHueDeviceController(HueDeviceController hueDeviceController) {
        this.hueDeviceController = hueDeviceController;
    }
}

File: src/main/node/com/smarthome/backend/server/api/modules/hue/devices/hueTemperatureSensor.ts
import { logger } from "../../../../../logger.js";
import { DeviceTemperature } from "../../../../../model/devices/DeviceTemperature.js";
import { HueDeviceController } from "../hueDeviceController.js";

export class HueTemperatureSensor extends DeviceTemperature {
  private bridgeId?: string;
  private hueResourceId?: string;
  private batteryRid?: string;
  private hueDeviceController?: HueDeviceController;

  constructor(
    name?: string,
    id?: string,
    bridgeId?: string,
    hueResourceId?: string,
    batteryRid?: string,
    hueDeviceController?: HueDeviceController
  ) {
    super();
    if (name) this.name = name;
    if (id) this.id = id;
    this.bridgeId = bridgeId;
    this.hueResourceId = hueResourceId;
    this.batteryRid = batteryRid;
    this.hueDeviceController = hueDeviceController;
    this.moduleId = "hue";
    this.isConnected = true;
  }

  async updateValues(): Promise<void> {
    if (!this.hueDeviceController || !this.bridgeId || !this.hueResourceId) {
      logger.debug(
        { deviceId: this.id, hasController: !!this.hueDeviceController, bridgeId: this.bridgeId, hueResourceId: this.hueResourceId },
        "HueTemperatureSensor.updateValues uebersprungen"
      );
      return;
    }
    const status = await this.hueDeviceController.getTemperature(this.bridgeId, this.hueResourceId);
    if (!status || typeof status.temperature !== "number") return;
    const rounded = Math.round(status.temperature);
    if (this.temperature !== rounded) {
      await this.setTemperature(rounded, false);
    }
  }

  protected async executeSetTemperature(temperature: number): Promise<void> {
    return;
  }

  getBridgeId() {
    return this.bridgeId;
  }

  setBridgeId(bridgeId: string) {
    this.bridgeId = bridgeId;
  }

  getHueResourceId() {
    return this.hueResourceId;
  }

  setHueResourceId(hueResourceId: string) {
    this.hueResourceId = hueResourceId;
  }

  getBatteryRid() {
    return this.batteryRid;
  }

  setBatteryRid(batteryRid: string) {
    this.batteryRid = batteryRid;
  }

  setHueDeviceController(hueDeviceController: HueDeviceController) {
    this.hueDeviceController = hueDeviceController;
  }
}

